tests/apollo/test_skvbc_reconfiguration.py

```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
+        for i in range(100):
+            await skvbc.write_known_kv()
+
+        for r in bft_network.all_replicas():
+            assert (r < 4)
+            num_of_fast_path = await self._get_metric(r, bft_network, "Counters", "totalFastPaths")
+            self.assertGreater(num_of_fast_path, 0)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 7)
+    async def test_remove_nodes_with_f_failures(self, bft_network):
+        """
+        In this test we show how a system operator can remove nodes (and thus reduce the cluster) from 7 nodes cluster
+        to 4 nodes cluster even when f nodes are not responding
+        For that the operator performs the following steps:
+        1. Stop 2 nodes (f=2)
+        2. Send a remove_node command - this command also wedges the system
+        3. Verify that all live (including the removed candidates) have stopped
+        4. Load  a new configuration to the bft network
+        5. Rerun the cluster with only 4 nodes and make sure they succeed to perform transactions in fast path
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        client = bft_network.random_client()
+
+        for i in range(100):
+            await skvbc.write_known_kv()
+        # choose two replicas to crash and crash them
+        crashed_replicas = {5, 6} # For simplicity, we crash the last two replicas
+        bft_network.stop_replicas(crashed_replicas)
+
+        num_of_slow_path_before = await self._get_metric(0, bft_network, "Counters", "totalSlowPaths")
+        # All next request should be go through the slow path
+        for i in range(100):
+            await skvbc.write_known_kv()
+
+        key, val = await skvbc.write_known_kv()
+
+        live_replicas = bft_network.all_replicas(without=crashed_replicas)
+        # Verify that replicas are now in slow path
+        for r in live_replicas:
+            num_of_slow_path = await self._get_metric(r, bft_network, "Counters", "totalSlowPaths")
+            self.assertGreater(num_of_slow_path, num_of_slow_path_before)
+
+        await client.write(skvbc.write_req([], [], block_id=0, add_remove_node_command=True))
+
+        with trio.fail_after(seconds=90):
+            done = False
+            while done is False:
+                msg = skvbc.get_have_you_stopped_req(0)
+                rep = await client.read(msg, m_of_n_quorum=bft_client.MofNQuorum(live_replicas, len(live_replicas)))
+                rsi_rep = client.get_rsi_replies()
+                done = True
+                for r in rsi_rep.values():
+                    if skvbc.parse_rsi_reply(rep, r) == 0:
+                        done = False
+                        break
+
+        for r in live_replicas:
+            last_stable_checkpoint = await self._get_metric(r, bft_network, "Gauges", "lastStableSeqNum")
+            self.assertGreaterEqual(last_stable_checkpoint, 300)
```

Comment:
where does the `300` come from?

Reply:
This is the next next checkpoint (the checkpoint we want to stop at).
Anyway, I have added a variable for this